cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -83,40 +87,151 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
+        private final Service service;
+        private final Service headlessService;
+        private final ConfigMap metricsConfigMap;
+        private final StatefulSet statefulSet;
+        private final Secret clientsCASecret;
+        private final Secret clientsPublicKeySecret;
+        private final Secret brokersClientsSecret;
+        private final Secret brokersInternalSecret;
+        private ReconcileResult<StatefulSet> diffs;
+
+        KafkaClusterDescription(KafkaCluster kafka, Service service, Service headlessService,
+                                ConfigMap metricsConfigMap, StatefulSet statefulSet,
+                                Secret clientsCASecret, Secret clientsPublicKeySecret,
+                                Secret brokersClientsSecret, Secret brokersInternalSecret) {
+            this.kafka = kafka;
+            this.service = service;
+            this.headlessService = headlessService;
+            this.metricsConfigMap = metricsConfigMap;
+            this.statefulSet = statefulSet;
+            this.clientsCASecret = clientsCASecret;
+            this.clientsPublicKeySecret = clientsPublicKeySecret;
+            this.brokersClientsSecret = brokersClientsSecret;
+            this.brokersInternalSecret = brokersInternalSecret;
+        }
+
+        KafkaCluster kafka() {
+            return this.kafka;
+        }
+
+        Service service() {
+            return this.service;
+        }
+
+        Service headlessService() {
+            return this.headlessService;
+        }
+
+        ConfigMap metricsConfigMap() {
+            return this.metricsConfigMap;
+        }
+
+        StatefulSet statefulSet() {
+            return this.statefulSet;
+        }
+
+        Secret clientsCASecret() {
+            return this.clientsCASecret;
+        }
+
+        Secret clientsPublicKeySecret() {
+            return this.clientsPublicKeySecret;
+        }
+
+        Secret brokersClientsSecret() {
+            return this.brokersClientsSecret;
+        }
+
+        Secret brokersInternalSecret() {
+            return this.brokersInternalSecret;
+        }
+
+        ReconcileResult<StatefulSet> diffs() {
+            return this.diffs;
+        }
+
+        Future<KafkaClusterDescription> withDiff(Future<ReconcileResult<StatefulSet>> r) {
+            return r.map(rr -> {
+               this.diffs = rr;
+               return this;
+            });
+        }
+
+        Future<KafkaClusterDescription> withVoid(Future<?> r) {
+            return r.map(this);
+        }
+    }
+
+    private final Future<KafkaClusterDescription> getKafkaCluster(ConfigMap assemblyCm, List<Secret> assemblySecrets) {
+        Future<KafkaClusterDescription> fut = Future.future();
+
+        vertx.createSharedWorkerExecutor("kubernetes-ops-pool").executeBlocking(
+                future -> {
+                    try {
+                        KafkaCluster kafka = KafkaCluster.fromDescription(assemblyCm, assemblySecrets);
```

Comment:
Again, this feels strange. You called this method `fromDescription`. From that name my expectation would be that it would accept the `Description` class. But it seems to accept just regular ConfigMap and list of Secrets. And the `Description` class is created only from the result of `fromDescription` method That doesn't make sense. IMHO can rename `fromDescription` back to `fromConfigMap`

Reply:
But `fromConfigMap` sounds to expect just a `ConfigMap`. Maybe we should think about a better name?

Reply:
I'm fine with something better if you have it. But `fromDescription` sounds confusing when you actually use a class called `*Description` which is not used here. The ConfigMap is at least still the main source of information.

Reply:
At this point I agree on revert to `fromConfigMap`; it will have less impact on merge and waiting for the CRD PR as well.

Reply:
We should keep a list somewhere of all the tidy up tasks we have between the encryption stuff and the crd stuff, so we actually do it after they're all merged.